### PR TITLE
Improvements to the context management

### PR DESCRIPTION
**Pull request by niwinz**
This is a initial work to improve the context management with first goal make the context management more decoupled from monad concept.

TODO:
- [x] Reimplement the context management making it less coupled to monad and monad transformer.
- [x] Make the with-context macro more strict, only allowing a valid context objects.
- [x] Rename all `stuff-monad` to `context` or `stuff-context`.
- [x] Add constants for context class level
- [x] Rename al protocol methods to be in the `-method` style name.
- [x] Revisit some TODOS
- [x] Revisit the state/writer/reader implementation and its api.

### CODE REVIEW

**Comment by niwinz** [comment]
I'm not pretty sure about that you are saying with automate satisfying protocols. Could you explain it a little bit more?

**Comment by yurrriq** [comment]
Wow, I typed that on my phone and the auto-correct was rather unhelpful..

Anyway, I just noticed a lot of nearly identical code between the different Monad instances.

For example, I've copied the definitions [`Right`][1], [`Left`][2], and [`Just`][3] to the files, `right.clj`, `left.clj` and `just.clj`, respectively.

```bash
$ wc -l right.clj left.clj just.clj
```

```
      27 right.clj
      27 left.clj
      27 just.clj
      81 total
```

These three snippets are 27 lines each, or 81 lines total. Between them, though, there only a few subtle differences. `Nothing` differs a bit more, but still not by very much.

```bash
$ diff3 right.clj left.clj just.clj
```

```diff
====
1:1c
  (deftype Right [v]
2:1c
  (deftype Left [v]
3:1c
  (deftype Just [v]
====
1:15c
           (if (instance? Right other)
2:15c
           (if (instance? Left other)
3:15c
           (if (instance? Just other)
====
1:25c
           (if (instance? Right other)
2:25c
           (if (instance? Left other)
3:25c
           (if (instance? Just other)
```

It seems to me there must be a good way to cut down on some of this boilerplate code, perhaps my introducing a `defmonad` macro that, given the correct parameters, generates the code I linked above, allowing for customizations (e.g. `Nothing`) as appropriate.

Does that make any more sense?


[1]: https://github.com/funcool/cats/blob/1894523212fbac9ffd608f0e6353cf7ffdd7a830/src/cats/monad/either.cljc#L48-L74
[2]: https://github.com/funcool/cats/blob/1894523212fbac9ffd608f0e6353cf7ffdd7a830/src/cats/monad/either.cljc#L76-L102
[3]: https://github.com/funcool/cats/blob/1894523212fbac9ffd608f0e6353cf7ffdd7a830/src/cats/monad/maybe.cljc#L43-L69

**Comment by niwinz** [comment]
Now I understand @yurrriq , thanks for the explanation. Personally I don't like abuse of macros. Use macros obviously will reduce some boilerplate code but it will make it less accessible to one that will want undertand the cats code (forcing understand the specific domain macros used internally for define monads). 

Until now, we have used plain clojure constructions for make it more simple and direct with cost of some boilerplate code for define some predicates and more semantic constructors (the defrecord also defines ones but them are looks ugly... ;p).

Obviously, use macros will allow us do some more smart thinks like define conditional implementations in monad transformers and is a cool feature but at this moment I can't valuate the tradeoff of user o not use the macros. Maybe in a future we can think in some special macro for define transformers giving them a little bit of power but for generic macros for define monads at this moment I don't have any convincing argument for ti.

**Comment by niwinz** [comment]
@dialelo please give me sone feedback please about changes introduced in this pr..

**Comment by yurrriq** [comment]
Cool. I appreciate your reasoning and explanation. Ok second thought, it would be a total abuse of macros.

File: src/cats/applicative/validation.cljc [hunk 1]
```diff
@@ -131,53 +131,56 @@
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
 (def ^{:no-doc true}
-  validation-applicative
+  context
   (reify
+    p/ContextClass
+    (-get-level [_] 10)
```
Comment (dialelo) [inline src/cats/applicative/validation.cljc]: We may want to use the constant here.

**Comment by dialelo** [comment]
Excellent work, I'm merging this.

**Pull request merged**